src/Microsoft.AspNet.JsonPatch/Helpers/ObjectTreeAnalyisResult.cs

```diff
@@ -0,0 +1,211 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Newtonsoft.Json.Serialization;
+
+namespace Microsoft.AspNet.JsonPatch.Helpers
+{
+    internal class ObjectTreeAnalysisResult
+    {
+        // either the property is part of the container dictionary,
+        // or we have a direct reference to a JsonPatchProperty instance
+
+        public bool UseDynamicLogic { get; private set; }
+
+        public bool IsValidPathForAdd { get; private set; }
+
+        public bool IsValidPathForRemove { get; private set; }
+
+        public IDictionary<string, object> Container { get; private set; }
+
+        public string PropertyPathInParent { get; private set; }
+
+        public JsonPatchProperty JsonPatchProperty { get; private set; }
+
+        public ObjectTreeAnalysisResult(object objectToSearch, string propertyPath
+            , IContractResolver contractResolver)
+        {
+            // construct the analysis result.
+
+            // split the propertypath, and if necessary, remove the first 
+            // empty item (that's the case when it starts with a "/")
+
+            var propertyPathTree = propertyPath.Split('/').ToList();
+            object targetObject = objectToSearch;
+
+            if (string.IsNullOrWhiteSpace(propertyPathTree[0]))
+            {
+                // remove it
+                propertyPathTree.RemoveAt(0);
+            }
+
+            // we've now got a split up property tree "base/property/otherproperty/..."
+            int lastPosition = 0;
+            for (int i = 0; i < propertyPathTree.Count; i++)
+            {
+                // if the current target object is an ExpandoObject (IDictionary<string, object>),
+                // we cannot use the ContractResolver.
```

Comment:
I would think that it would return `JsonDynamicContract` here, which should do the right thing wrt property name lookups, and allow you to add a property if it doesn't exist. Is that not the case?